source/extensions/matching/common_inputs/transport_socket/config.h

```diff
@@ -0,0 +1,199 @@
+  createDataInputFactoryCb(const Protobuf::Message& config,
+                           ProtobufMessage::ValidationVisitor& validation_visitor) override;
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+
+  std::string category() const override { return "envoy.matching.inputs"; }
+};
+
+/**
+ * Data input for extracting values from filter state.
+ * This enables downstream-connection-based matching by reading filter state that was
+ * explicitly shared from downstream to upstream via TransportSocketOptions.
+ */
+class FilterStateInput : public TransportSocketInputBase {
+public:
+  explicit FilterStateInput(const std::string& key) : key_(key) {}
+
+protected:
+  absl::optional<std::string> getValue(const TransportSocketMatchingData& data) const override;
+
+private:
+  const std::string key_;
+};
+
+/**
+ * Factory for creating filter state data inputs.
+ */
+class FilterStateInputFactory : public Matcher::DataInputFactory<TransportSocketMatchingData> {
+public:
+  std::string name() const override {
+    return "envoy.matching.inputs.transport_socket_filter_state";
+  }
+
+  Matcher::DataInputFactoryCb<TransportSocketMatchingData>
+  createDataInputFactoryCb(const Protobuf::Message& config,
+                           ProtobufMessage::ValidationVisitor& validation_visitor) override;
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+
+  std::string category() const override { return "envoy.matching.inputs"; }
+};
+
+/**
+ * Action that carries a transport socket name.
+ */
+class TransportSocketNameAction : public Matcher::Action {
+public:
+  explicit TransportSocketNameAction(const std::string& name) : name_(name) {}
+  const std::string& name() const { return name_; }
+  absl::string_view typeUrl() const override {
+    return "type.googleapis.com/"
+           "envoy.extensions.matching.common_inputs.transport_socket.v3.TransportSocketNameAction";
+  }
+
+private:
+  const std::string name_;
+};
+
+/**
+ * ActionFactory that creates TransportSocketNameAction.
+ */
+class TransportSocketNameActionFactory
+    : public Matcher::ActionFactory<Server::Configuration::ServerFactoryContext> {
+public:
+  std::string name() const override { return "envoy.matching.action.transport_socket.name"; }
+  Matcher::ActionConstSharedPtr createAction(const Protobuf::Message& config,
+                                             Server::Configuration::ServerFactoryContext&,
+                                             ProtobufMessage::ValidationVisitor&) override;
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+};
+
+} // namespace TransportSocket
+} // namespace CommonInputs
+} // namespace Matching
+} // namespace Extensions
+
+// Export TransportSocketMatchingData to the Upstream namespace for backward compatibility.
+namespace Upstream {
+using TransportSocketMatchingData =
+    Extensions::Matching::CommonInputs::TransportSocket::TransportSocketMatchingData;
```

Comment:
Why not move this into `envoy/`? I think that would break the dependency problem of core depending on the extension?